packages/realm/src/Counter.ts

```diff
@@ -0,0 +1,148 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2024 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+import { IllegalConstructorError, Realm, UpdateMode, assert, binding } from "./internal";
+
+const REALM = Symbol("Counter#realm");
+const OBJ = Symbol("Counter#obj");
+const COLUMN_KEY = Symbol("Counter#columnKey");
+
+/**
+ * A logical counter representation for performing numeric updates that need
+ * to be synchronized as sequentially consistent events rather than individual
+ * reassignments of the number.
+ *
+ * For instance, offline Client 1 and Client 2 which both see `Counter.value`
+ * as `0`, can both call `Counter.increment(1)`. Once online, the value will
+ * converge to `2`.
+ *
+ * ### Counter types are *not* supported as:
+ *
+ * - `Mixed` values
+ * - Primary keys
+ * - Inside collections
+ * - Query arguments for placeholders (e.g. `$0`) in {@link Realm.Results.filtered | filtered()}
+ *   - If you need to use the value of the `Counter` when filtering, use `Counter.value`.
+ *
+ * ### Declaring a counter
+ *
+ * A property schema is declared as either:
+ * - `"counter"`
+ * - `{ type: "int", presentation: "counter" }`
+ *
+ * ### Creating a counter
+ *
+ * Use a `number` when creating your counter on a {@link Realm.Object}.
+ * ```ts
+ * realm.write(() => {
+ *   realm.create(MyObject, { _id: "123", counter: 0 });
+ * });
+ * ```
+ *
+ * To update a `null` counter, use {@link UpdateMode.Modified} or {@link UpdateMode.All}.
+ * ```ts
+ * realm.write(() => {
+ *   realm.create(MyObject, { _id: "123", counter: 0 }, UpdateMode.Modified);
+ * });
+ * ```
+ *
+ * ### Nullability
+ *
+ * The above property schema can be extended to allow a nullable counter.
+ * A `Counter` never stores `null` values itself, but the counter property
+ * on the {@link Realm.Object} (e.g. `myRealmObject.myCounter`) can be `null`.
+ *
+ * ### Not a live object
+ *
+ * The `Counter` instance itself is *not* a live object. Calling `myCounter.value`
+ * will always return the most recent locally persisted (and synced) state, and
```

Comment:
> the most recent locally persisted

That makes it live, doesn't it?

Reply:
The distinction I made here between the Counter itself and its value may not be correct in terms of "live", thanks for pointing it out 🙂